lib/web_ui/lib/src/engine/canvaskit/image_web_codecs.dart

```diff
@@ -0,0 +1,436 @@
+      preferAnimation: true,
+    ));
+
+    try {
+      await js_util.promiseToFuture<void>(webDecoder.tracks.ready);
+
+      // Flutter doesn't have an API for progressive loading of images, so we
+      // wait until the image is fully decoded.
+      // package:js bindings don't work with getters that return a Promise, which
+      // is why js_util is used instead.
+      await js_util.promiseToFuture<void>(js_util.getProperty(webDecoder, 'completed'));
+    } catch (error) {
+      if (error is html.DomException) {
+        if (error.name == html.DomException.NOT_SUPPORTED) {
+          throw ImageCodecException(
+            'Image file format ($contentType) is not supported by this browser\'s ImageDecoder API.\n'
+            'Image source: $debugSource',
+          );
+        }
+      }
+      throw ImageCodecException(
+        'Failed to decode image using the browser\'s ImageDecoder API.\n'
+        'Image source: $debugSource\n'
+        'Original browser error: $error'
+      );
+    }
+    return CkBrowserImageDecoder._(webDecoder, debugSource);
+  }
+
+  CkBrowserImageDecoder._(this.webDecoder, this.debugSource);
+
+  final _ImageDecoder webDecoder;
+  final String debugSource;
+
+  /// Whether this decoded has been disposed of.
+  ///
+  /// Once this turns true it stays true forever, and this decoder becomes
+  /// unusable.
+  bool _isDisposed = false;
+
+  @override
+  void dispose() {
+    _isDisposed = true;
+
+    // This releases all resources, including any currently running decoding work.
+    webDecoder.close();
+  }
+
+  void _debugCheckNotDisposed() {
+    assert(
+      !_isDisposed,
+      'Cannot use this image decoder. It has been disposed of.'
+    );
+  }
+
+  @override
+  int get frameCount {
+    _debugCheckNotDisposed();
+    return webDecoder.tracks.selectedTrack!.frameCount;
+  }
+
+  /// The index of the frame that will be decoded on the next call of [getNextFrame];
+  int _nextFrameIndex = 0;
+
+  @override
+  Future<ui.FrameInfo> getNextFrame() async {
+    _debugCheckNotDisposed();
+    final _DecodeResult result = await js_util.promiseToFuture<_DecodeResult>(
+      webDecoder.decode(_DecodeOptions(frameIndex: _nextFrameIndex)),
+    );
+    final _VideoFrame frame = result.image;
+    _nextFrameIndex = (_nextFrameIndex + 1) % frameCount;
+
+    // A surface is required because the image is uploaded directly to GPU, so
+    // we need a WebGL context. However, it is possible that the image is being
+    // decoded before the first frame is rendered. This will create the base
+    // surface, if necessary, or reuse the existing one.
+    final CkSurface ckSurface = SurfaceFactory.instance.baseSurface.createOrUpdateSurface(ui.window.physicalSize);
+    final SkSurface skSurface = ckSurface.surface;
+    final SkImage? skImage = skSurface.makeImageFromTextureSource(
+      frame,
+      frame.displayWidth,
+      frame.displayHeight,
+    );
+
```

Comment:
If you're done with frame at this point it's good practice to call frame.close() to free up memory.

Reply:
Done.